text/0030-shared-event-queue.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: Shared ThreadsafeFunction in EventQueue
+- Start Date: 2021-05-19
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Make `EventQueue` use a shared `ThreadsafeFunction` to leverage the
+[performance optimization][0] in recent Node.js versions.
+
+# Motivation
+[motivation]: #motivation
+
+`EventQueue` provides a way to invoke Rust closures on a JavaScript thread from
+another thread by using a `ThreadsafeFunction`. Recent Node.js versions have a
+[performance optimization][0] for the case where a single `ThreadsafeFunction`
+is invoked multiple times in a row. Thus using a single shared
+`ThreadsafeFunction` in `EventQueue`s would lead to significant performance
+improvements for Neon users.
+
+These improvements are perhaps most noticeable when Neon addon is used from the
+Electron, because scheduling new UV ticks in Electron are very costly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Whenever `cx.queue()` is invoked - a new instance of `EventQueue` is created.
+With ithe proposed changes the instance would point to a shared
+`ThreadsafeFunction` internally. For the most scenarios there would be no
+difference between current and proposed implementation. Calling `queue.send()`
+will still invoke the closure on the JavaScript thread:
+```rust
+let queue = cx.queue();
+
+std::thread::spawn(move || {
+    queue.send(|| {
+        Ok(())
+    })
+});
+```
+
+Calling `queue.send()` several times in a row using different `cx.queue()` would
+be significantly faster:
+```rust
+loop {
+    cx.queue().send(|| {
+        Ok(())
+    })
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+With the recent [node.js changes][0] `ThreadsafeFunction` will loop over pending
+calls, invoking up to 1000 of them in the same tick instead of running each new
+call on a separate UV loop tick. This is an implementation detail that is
+intentionally not covered by `napi_create_threadsafe_function` documentation to
+allow for potential performance optimizations in the future.
+
+Neon doesn't currently leverage it without the proposed changes because a new
```

Comment:
Neon _does_ leverage the change, but it requires the user to continue to re-use the same queue. This is one of the reasons queues are `Sync`.

I think it's important to note that this is an API change with the intent to guide users towards the choice that is optimal in most situations.

Reply:
> @indutny Thanks for providing this. I think this is a really interesting idea.
> 
> One complexity concern I have is how dropping a referenced `EventQueue` would work. With a single shared queue, the queue needs to track how many times it has been referenced. When the count drops to `0`, the shared queue needs to unreference itself.
> 
> However, unreferencing can only be performed on the event loop. The only solution I can think of is sending an _event_ to unreference itself. This could end up being costly if referenced queues are frequently created and dropped.

This is indeed the case, but I don't believe that it will cause any performance losses because of how fast the threadsafe functions are now. I've added an extra optimization to the [PR to neon](https://github.com/neon-bindings/neon/pull/739) to avoid this extra work when `EventQueue` owns the `ThreadsafeFunction` and will drop it.